File: java/leetcode_grind/Day450.java
package leetcode_grind;

import java.util.HashMap;
import java.util.List;
import java.util.PriorityQueue;

public class Day450 {
    // https://leetcode.com/problems/smallest-range-covering-elements-from-k-lists/description/
    static class Solution1 {
        public int[] smallestRange(List<List<Integer>> nums) {
            int n = nums.size();

            int minx = 0;
            int miny = Integer.MAX_VALUE;
            int max = Integer.MIN_VALUE;

            int[] next = new int[n];

            PriorityQueue<Integer> pq = new PriorityQueue<>(
                    (a, b) -> nums.get(a).get(next[a]) - nums.get(b).get(next[b]));
            for (int i = 0; i < n; ++i) {
                pq.offer(i);
                max = Math.max(max, nums.get(i).get(0));
            }

            outer: for (int i = 0; i < n; ++i) {
                for (int j = 0; j < nums.get(i).size(); ++j) {
                    int mini = pq.poll();

                    if (max - nums.get(mini).get(next[mini]) < miny - minx) {
                        minx = nums.get(mini).get(next[mini]);
                        miny = max;
                    }

                    next[mini] += 1;
                    if (next[mini] == nums.get(mini).size()) {
                        break outer;
                    }

                    pq.offer(mini);
                    max = Math.max(max, nums.get(mini).get(next[mini]));
                }
            }

            return new int[] { minx, miny };
        }
    }

    // https://leetcode.com/problems/first-unique-character-in-a-string/description/
    static class Solution2 {
        public int firstUniqChar(String s) {
            var map = new HashMap<Character, Integer>();
            for (int i = 0; i < s.length(); ++i) {
                map.put(s.charAt(i), map.getOrDefault(s.charAt(i), 0) + 1);
            }
            int idx = -1;
            for (int i = 0; i < s.length(); ++i) {
                if (map.get(s.charAt(i)) == 1) {
                    return i;
                }
            }
            return idx;
        }
    }
}

File: src/leetcode_grind/day_450.rs
// https://leetcode.com/problems/circular-array-loop/description/
pub fn circular_array_loop(nums: Vec<i32>) -> bool {
    let n = nums.len();
    if n == 1 {
        return false;
    }

    use std::collections::HashSet;
    let mut vis = HashSet::new();

    for k in 0..n {
        if vis.contains(&k) {
            continue;
        }

        let mut cycleset = HashSet::new();
        let mut i = k;
        loop {
            // println!("enter {}", i);
            if cycleset.contains(&i) {
                return true;
            }

            vis.insert(i);
            cycleset.insert(i);

            let prev = i;
            i = ((((i as i32 + nums[i]) % n as i32) + n as i32) % n as i32) as usize;
            // println!("{} {}", prev, i);

            if prev == i {
                // println!("prev == i");
                break;
            }

            if (nums[prev] > 0) == (nums[i] < 0) {
                // println!("{} > 0 != {} < 0", nums[prev], nums[i]);
                break;
            }
        }
    }
    false
}

// https://leetcode.com/problems/smallest-range-covering-elements-from-k-lists/description/
pub fn smallest_range(nums: Vec<Vec<i32>>) -> Vec<i32> {
    let mut next = vec![0; nums.len()];
    let (mut minx, mut miny) = (0, i32::MAX);

    'outer: for i in 0..nums.len() {
        for _ in 0..nums[i].len() {
            let (mut min, mut max) = (0, 0);

            for k in 0..next.len() {
                if nums[min][next[min]] > nums[k][next[k]] {
                    min = k;
                }
                if nums[max][next[max]] < nums[k][next[k]] {
                    max = k;
                }
            }

            if nums[max][next[max]] - nums[min][next[min]] < miny - minx {
                minx = nums[min][next[min]];
                miny = nums[max][next[max]];
            }

            next[min] += 1;

            if next[min] == nums[min].len() {
                break 'outer;
            }
        }
    }

    vec![minx, miny]
}

#[test]
fn test_circular_array() {
    let res = circular_array_loop(vec![-1, -3, 9]);
    println!("{res}");
}
